src/scheduler/scheduler.rs

```diff
@@ -51,80 +51,64 @@ unsafe impl<VM: VMBinding> Sync for GCWorkScheduler<VM> {}
+            WorkBucketStage::FinalRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::PhantomRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::CalculateForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::SecondRoots => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::RefForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::FinalizableForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Compact => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Release => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Final => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
         };
 
         // Set the open condition of each bucket.
         {
             // Unconstrained is always open. Prepare will be opened at the beginning of a GC.
             // This vec will grow for each stage we call with open_next()
-            let mut open_stages: Vec<WorkBucketStage> = vec![Unconstrained, Prepare];
+            let first_stw_stage = WorkBucketStage::first_stw_stage();
+            let mut open_stages: Vec<WorkBucketStage> = vec![first_stw_stage];
             // The rest will open after the previous stage is done.
-            let mut open_next = |s: WorkBucketStage| {
-                let cur_stages = open_stages.clone();
-                work_buckets[s].set_open_condition(move |scheduler: &GCWorkScheduler<VM>| {
-                    let should_open = scheduler.are_buckets_drained(&cur_stages)
-                        && scheduler.all_workers_parked();
-                    // Additional check before the `RefClosure` bucket opens.
-                    if should_open && s == crate::scheduler::work_bucket::LAST_CLOSURE_BUCKET {
-                        if let Some(closure_end) = scheduler.closure_end.lock().unwrap().as_ref() {
-                            if closure_end() {
-                                // Don't open `RefClosure` if `closure_end` added more works to `Closure`.
-                                return false;
+            let stages = (0..WorkBucketStage::LENGTH).map(WorkBucketStage::from_usize);
+            for stage in stages {
+                if stage != WorkBucketStage::Unconstrained && stage != first_stw_stage {
+                    let cur_stages = open_stages.clone();
+                    work_buckets[stage].set_open_condition(
+                        move |scheduler: &GCWorkScheduler<VM>| {
+                            let should_open = scheduler.are_buckets_drained(&cur_stages);
```

Comment:
The previous condition for opening a bucket is `are_buckets_drained() && all_workers_parked()`. But the current condition is simply `are_buckets_drained()`. So we may open a new bucket while some workers are executing packets from previous buckets (and they may generate more work for the previous bucket). I feel we should still make sure that all work is done by the workers. 

However, I changed this to `scheduler.are_buckets_drained() && scheduler.worker_group.all_parked()`. But it didn't fix the bug.

Reply:
The caller of this function assumes that all workers are parked. The `&& scheduler.worker_group.all_parked()` is redundant.

https://github.com/mmtk/mmtk-core/pull/507/files/8f7db595be9ed1b5ee3e4fa347a704d5f0ca6422#diff-561a0acf6a922ed5201f43a704c5a68bfc2431156ceb7b9bd8fa5041917f67c6R236

https://github.com/mmtk/mmtk-core/pull/507/files/8f7db595be9ed1b5ee3e4fa347a704d5f0ca6422#diff-561a0acf6a922ed5201f43a704c5a68bfc2431156ceb7b9bd8fa5041917f67c6R376-R377

Reply:
I see.